File: projects/shared-memory/reader.c
/*-------------map_normalfile2.c-----------*/
#include <sys/mman.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>

#define file "/tmp/shared-memory-file"
#define SHARED_SIZE 1024

int main(int argc, char** argv)  // map a normal file as shared mem:
{
  int fd,i;
  char *p;
  fd=open( file, O_CREAT|O_RDWR, 00777 );
  p = (char*)mmap(NULL, SHARED_SIZE ,PROT_READ|PROT_WRITE, MAP_SHARED,fd,0);
  for(i = 0;i<13;i++)
  {
    putchar(p[i]);
  }
  munmap( p,SHARED_SIZE );
}

File: shared-memory/reader.py
import mmap
import fcntl

file = "/tmp/shared-memory-file-123"

with open(file, "r+b") as f:
    mm = mmap.mmap(f.fileno(), 0)
    print 'wait lock'
    fcntl.flock(f, fcntl.LOCK_EX)
    print 'get lock'
    print mm.readline()
    fcntl.flock(f, fcntl.LOCK_UN)
    mm.close()